### PR TITLE
Make the constructor of Ref private. Fixes #951

### LINKED ISSUE #951: Public constructor of Ref allows breaking referential transparency
The expression `new Ref[IO, Int]` is not pure. Making the constructor private would force users to use the pure "equivalent": `Ref.uninitialized[IO, Int]`.

### PR DESCRIPTION
**Pull request by SystemFw**
Fixes #951

### CODE REVIEW

File: core/shared/src/main/scala/fs2/async/Ref.scala [hunk 1]
```diff
@@ -15,7 +15,7 @@ import java.util.concurrent.atomic.{AtomicBoolean,AtomicReference}
 import Ref._
 
 /** An asynchronous, concurrent mutable reference. */
-final class Ref[F[_],A](implicit F: Effect[F], ec: ExecutionContext) { self =>
+final class Ref[F[_],A] private[fs2] (implicit F: Effect[F], ec: ExecutionContext) { self =>
```
Comment (mpilquist) [inline core/shared/src/main/scala/fs2/async/Ref.scala]: We could make this `private` instead and change async.scala to use `Ref.uninitialized`

Comment (SystemFw) [reply core/shared/src/main/scala/fs2/async/Ref.scala]: That was my initial thought, then worried about the extra method call. Arguably over optimising though. (could also do `private[async]`).
Should I change it as you suggest?

Comment (SystemFw) [reply core/shared/src/main/scala/fs2/async/Ref.scala]: scratch it, `private[async]` won't work since it's used in Scope as well

Comment (mpilquist) [reply core/shared/src/main/scala/fs2/async/Ref.scala]: Ah right, merging then

**Pull request merged**